### ISSUE TITLE
Fatal error: uncaught exception File "backend/Regalloc.ml", line 249, characters 18-24: Assertion failed

### ISSUE THREAD

**Issue by zhendongsu** [AbsInt/CompCert#183]
```c
$ ccomp -version
The CompCert C verified compiler, version 3.0.1 (commit 7eaaa6a)
$ 
$ cat small.c
int printf (const char *, ...); 

int a; 
long long b[4][4][4];

int main ()
{
  int c = 300000000; 
  if (a) 
    printf ("%lld\n", b[a][a][c]);
  return 0; 
}
$ 
$ ccomp small.c
Fatal error: uncaught exception File "backend/Regalloc.ml", line 249, characters 18-24: Assertion failed
Raised at file "backend/Regalloc.ml", line 249, characters 18-30
Called from file "extraction/Coqlib.ml", line 68, characters 17-22
Called from file "extraction/Maps.ml", line 187, characters 40-56
Called from file "extraction/Maps.ml", line 187, characters 28-38
Called from file "extraction/Maps.ml", line 187, characters 28-38
Called from file "backend/Regalloc.ml", line 312, characters 11-75
Called from file "backend/Regalloc.ml", line 1191, characters 15-48
Called from file "extraction/Allocation.ml", line 1436, characters 11-21
Called from file "extraction/AST.ml", line 553, characters 9-25
Called from file "extraction/AST.ml", line 381, characters 12-27
Called from file "extraction/AST.ml", line 393, characters 15-55
Called from file "extraction/AST.ml", line 393, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 393, characters 15-55
Called from file "extraction/AST.ml", line 406, characters 8-57
Called from file "extraction/Compiler.ml", line 97, characters 18-1023
Called from file "driver/Driver.ml", line 53, characters 15-50
Called from file "driver/Driver.ml", line 75, characters 2-36
Called from file "driver/Driveraux.ml", line 137, characters 34-40
Called from file "driver/Driver.ml", line 484, characters 22-70
$ 
```

**Comment by bschommer** [comment]
It seems that only the x86 backend is affected.

**Comment by xavierleroy** [comment]
Only the x86-32 (IA32) back-end, actually.

The failure comes from the following RTL load instruction
```
	x4 = int64[x6 + 2400000000]
```
which fails the `Op.addressing_valid` test.

Should not be hard to fix, but I can't look into it right now, so please leave open.

**Comment by zhendongsu** [comment]
Here is a related, but distinct, crash with a similar store instruction: 

```c
$ cat small2.c           
int a; 
long long b[4][4][4];

int main ()
{
  int c = 300000000; 
  if (a) 
    b[a][a][c] = 0; 
  return 0; 
}
$ 
$ ccomp -O -fall small2.c
Fatal error: uncaught exception File "backend/Regalloc.ml", line 261, characters 18-24: Assertion failed
Raised at file "backend/Regalloc.ml", line 261, characters 18-30
Called from file "extraction/Coqlib.ml", line 68, characters 17-22
Called from file "extraction/Maps.ml", line 187, characters 40-56
Called from file "extraction/Maps.ml", line 187, characters 58-68
Called from file "backend/Regalloc.ml", line 312, characters 11-75
Called from file "backend/Regalloc.ml", line 1191, characters 15-48
Called from file "extraction/Allocation.ml", line 1436, characters 11-21
Called from file "extraction/AST.ml", line 553, characters 9-25
Called from file "extraction/AST.ml", line 381, characters 12-27
Called from file "extraction/AST.ml", line 393, characters 15-55
Called from file "extraction/AST.ml", line 393, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 383, characters 15-55
Called from file "extraction/AST.ml", line 406, characters 8-57
Called from file "extraction/Compiler.ml", line 97, characters 18-1023
Called from file "driver/Driver.ml", line 53, characters 15-50
Called from file "driver/Driver.ml", line 75, characters 2-36
Called from file "driver/Driveraux.ml", line 137, characters 34-40
Called from file "driver/Driver.ml", line 484, characters 22-70
$ 
```

**Comment by xavierleroy** [comment]
Fixed by commit 7a44249 .
